### ISSUE TITLE
IoU should ignore void class

### ISSUE THREAD

**Issue by bermanmaxim** [DrSleep/tensorflow-deeplab-resnet#10]
Hello, I think the provided code in `evaluate.py` does not ignore the `void` label in ground truth masks (which is mapped to the background in the 1D labels); therefore, the performance is underestimated compared to the results of VOC devkit. Here is the [link to the line ignoring void pixels in the VOC devkit](https://github.com/weiliu89/VOCdevkit/blob/master/VOCcode/VOCevalseg.m#L55).

**Comment by DrSleep** [comment]
Hello.
I don't think this is the case. The implementation of streaming_mean_iou in TF sequentially accumulates confusion matrix. Confusion matrix, in turn, has a parameter `num_classes`, which if given, guarantees that the size of the matrix will be `num_classes` x `num_classes` (otherwise, `num_classes` will be set to the one plus the maximum value in either predictions or labels: https://github.com/tensorflow/tensorflow/blob/99fe61a8a8f3dd41b4e1e4dedfc53b45f67e88a7/tensorflow/python/ops/confusion_matrix.py#L96).
Now, if we provide a tensor with some values exceeding `num_classes`, these values will be correctly ignored. E.g., as a toy example:
```python
    tf.contrib.metrics.confusion_matrix([1, 0, 2, 4, 2, 1], [1, 2, 4, 0, 4, 1], num_classes=3) ==>
        [[0 0 1]
         [0 2 0]
         [0 0 0]]
  ```

**Comment by bermanmaxim** [comment]
I agree, but in this case, if you look at [train-Deeplab/utils.py](https://github.com/martinkersner/train-DeepLab/blob/master/utils.py), shouldn't we add a "21" label for the "void" class in the pascal palette?

**Comment by bermanmaxim** [comment]
By the way, I don't think this usage of confusion_matrix is documented; to ignore values, the `weights` argument should be used instead (weight=None for some values...)
I indeed ran into problems with your toy example: using
```
import tensorflow as tf
with tf.device('/cpu:0'):
    metric = tf.contrib.metrics.confusion_matrix([1, 0, 2, 4, 2, 1], [1, 2, 4, 0, 4, 1], num_classes=3)
print(tf.Session().run(metric))
```
results in a segfault for me (the segfault disappears if I do not specify num_classes, or, strangely, if I don't pin `tf.device` to `/cpu:0`).

**Comment by DrSleep** [comment]
Yes, when visualising the 'void' label is mapped to (0, 0, 0). 

It looks like a bug with TF when choosing the device explicitly. I will evaluate the model using the weights parameter and let you know whether the results are changed.

**Comment by DrSleep** [comment]
I can confirm that with weights set the results are improved: from 76.8 mIoU to 80.1 mIoU on the validation set.
I have used the code below, and will update the evaluation script accordingly. 
Thank you for letting know about the issue!

```python
pred = tf.reshape(pred, [-1,])
gt = tf.reshape(label_batch, [-1,])
weights = tf.cast(tf.less_equal(gt, 20), tf.int32)

# mIoU
mIoU, update_op = tf.contrib.metrics.streaming_mean_iou(pred, gt, num_classes=21, weights=weights)
```

**Comment by bermanmaxim** [comment]
Great. By the way, I think the evaluation codes in train_Deeplab have the same problem. I'll notify there too.

**Comment by DrSleep** [comment]
I have pushed the changes to the master branch. 
Thanks again!
